### PR TITLE
[OPIK-446] Add duration to trace and span

### PR DESCRIPTION
**Pull request by thiagohora**
## Details
- Add durations field to traces and spans
- Add filter by duration in our filter api

The duration was presented as double in milliseconds, but preserving the decimal to allow sub-milisecond precision

## Issues
OPIK-446

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/utils/DurationUtils.java
Comment (BorisTkachenko) [inline apps/opik-backend/src/main/java/com/comet/opik/utils/DurationUtils.java]: Its in milliseconds, not seconds

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java
Comment (BorisTkachenko) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: Why do we need to have `Instant start` as parameter? It's always the same and can be just instantiated within the test.
And for the end you could just pass span duration as long (1L for example) and do conversion in the test: `Duration.ofMillis(1L).plusNanos(1).toNanos() / 1000`. Will make arguments much shorter and more readable.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java
Comment (BorisTkachenko) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java]: Same as above

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java
Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: Good point!

**Review by BorisTkachenko — APPROVED** [review]

File: apps/opik-backend/src/main/java/com/comet/opik/api/Span.java
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: Why aren't you calculating in ClickHouse? You can do it on the flight or add a calculated field.

Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: Yeah, we can. I just thought it would be simple to calculate here since we can use the values. We can add a materialized column if you think it would be a better solution

Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: @andrescrz done in https://github.com/comet-ml/opik/pull/889

Comment (andrescrz) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: The important thing for the filters is to work, is that it's done in ClickHouse. It can be on the flight or as materialised column.

**Review by andrescrz — CHANGES_REQUESTED** [review]
Let's discuss a moment before moving forward. I want to make sure no issues.

File: apps/opik-backend/src/main/java/com/comet/opik/api/Span.java
Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: The filter was tested as well, but please feel free to look over the other PR. I implement the materialised column

Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: @andrescrz, can you look over the other pr where the materialized column is implemented? Can be found here: https://github.com/comet-ml/opik/pull/889/files#diff-97663d5cce992b8c2a80744f1b4c11c569bef17543914f1da68659b50a359add
I'm taking over this one as @thiagohora is out for a long vacation, if you think that's OK I would like to merge.

Comment (andrescrz) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: Let's nuke this service level calculation and just add a duration field in this `span` object. 

The actual calculation should be done on the query level.

File: apps/opik-backend/src/main/java/com/comet/opik/api/Trace.java
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/api/Trace.java]: Same. Let's nuke this an add a `duration` field.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/filter/FilterQueryBuilder.java
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/filter/FilterQueryBuilder.java]: This should be moved to the query level on the DAO, for both spans and traces. 

The reason is for a good separation of responsibilities, as filters shouldn't be calculating data. They should just take care of filtering.

Additionally, it isn't clear why dividing by 1000. It's inconsistent with the existing calculation code. Please se existing code as reference, example:

`if(end_time IS NOT NULL, date_diff('microsecond', start_time, end_time), null) as duration`

File: apps/opik-backend/src/main/java/com/comet/opik/utils/DurationUtils.java
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/utils/DurationUtils.java]: This should be nuked as well.

File: apps/opik-backend/src/main/java/com/comet/opik/api/Span.java
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: It seems that we're using durations as floating number instead of using ISO 8601 format. 

Although while this is inconsistent with our timestamp formatting, which uses ISO 8601, in the rest of the service, I don't have strong feelings against it. 

However, please double check that product is fine with this and specially frontend, as they might need to convert it to a more human readable format such as ISO 8601.

I'm not sure what the FE design is going to show here.

**Review by andrescrz — CHANGES_REQUESTED** [review]
I left my guidelines to move forward with this PR. 

Let's wait for the PRs on top of this until this is merged. 

Better go one by one to reduce noise and unnecessary efforts.

File: apps/opik-backend/src/main/java/com/comet/opik/api/Span.java
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: @andrescrz, product and FE agree that double millisecond durations are accepted

File: apps/opik-backend/src/main/java/com/comet/opik/domain/filter/FilterQueryBuilder.java
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/filter/FilterQueryBuilder.java]: Agree. I will use some of @thiagohora's changes in the later PR to use a materialized column instead.
As for the division by 1000, I believe the idea is to have a millisecond value with microsecond precision. If `date_diff('millisecond', ...)` is used than the result will be `0` for durations smaller than a millisecond.
As for why microsecond is used and not nanosecond, from what I see, `date_diff` support for `nanosecond` is not supported in the version we are using. [Support for this introduced](https://github.com/ClickHouse/ClickHouse/commit/5020741b253f66967ef84707c98a2fd201d5080d) several months ago.

File: apps/opik-backend/src/main/java/com/comet/opik/api/Span.java
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Span.java]: Done for spans in https://github.com/comet-ml/opik/pull/883/commits/af017e0022f558f0a9ef2c6b9afbf9ab301eadca. Will do something similar for traces.

File: apps/opik-backend/src/main/java/com/comet/opik/api/Trace.java
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/api/Trace.java]: Done in https://github.com/comet-ml/opik/pull/883/commits/177b12f045a9538b7424d1b908ee529b1001f075

File: apps/opik-backend/src/main/java/com/comet/opik/utils/DurationUtils.java
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/utils/DurationUtils.java]: Done in https://github.com/comet-ml/opik/pull/883/commits/00e0752e8c158086e97533a30a7a24620972212d

File: apps/opik-backend/src/main/java/com/comet/opik/domain/filter/FilterQueryBuilder.java
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/filter/FilterQueryBuilder.java]: Done in https://github.com/comet-ml/opik/pull/883/commits/d7ad75042780b89ed58b8b768e187423e2f38cca

**Review by andrescrz — APPROVED** [review]

File: apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000009_add_duration_columns.sql
Comment (andrescrz) [inline apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000009_add_duration_columns.sql:16]: Forgot to mention, that my initial comment was misunderstood. 

I mentioned that the calculation had to be moved from the backend service side to the ClickHouse query time, not necessarily that we had to add a calculated field.

This works fine from the functional perspective, but this migration might cause issues when deploying to an environment where spans or traces tables are big, as the new field would need to be calculated for all rows.

We're still on time to reconsider this.

Comment (idoberko2) [reply apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000009_add_duration_columns.sql:16]: Do you believe switching to a virtual column is a good alternative?

Comment (idoberko2) [reply apps/opik-backend/src/main/resources/liquibase/db-app-analytics/migrations/000009_add_duration_columns.sql:16]: Addressed in #916

**Pull request merged**
